# XML, HTML, and Web Scraping

JSON and XML are two different ways to represent hierarchical data. Which one is better? There are lots of articles online which discuss similarities and differences between JSON and XML and their advantages and disadvantages. Both formats are still in current usage, so it is good to be familiar with both. However, JSON is more common, so we'll focus on working with JSON representations of hierarchical data.

The reading covered an example of using Beautiful Soup to parse XML. Rather than doing another example XML now, we'll skip straight to scraping HTML from a webpage. Both HTML and XML can be parsed in a similar way with Beautiful Soup.

In [1]:
#!pip install beautifulsoup4
import pandas as pd

zsh:1: command not found: pip
zsh:1: command not found: pip


In [2]:
import requests

url = "https://tasty.p.rapidapi.com/recipes/list"

querystring = {"from":"0","size":"20","q":"daikon"}

headers = {
    "X-RapidAPI-Key": 'bcd546fbbcmsh57184bd607df5c9p15db5djsn46b87fb5f5b9',
    "X-RapidAPI-Host": "tasty.p.rapidapi.com"
}

response = requests.get(url, headers=headers, params=querystring)

print(response.json())

{'count': 13, 'results': [{'approved_at': 1570199933, 'aspect_ratio': '1:1', 'beauty_url': None, 'brand': None, 'brand_id': None, 'buzz_id': None, 'canonical_id': 'recipe:5643', 'compilations': [{'approved_at': 1607958492, 'aspect_ratio': '1:1', 'beauty_url': None, 'buzz_id': None, 'canonical_id': 'compilation:1893', 'country': 'US', 'created_at': 1606970945, 'description': 'Let your taste buds travel around the world with a new breakfast from a different country for each day of the week. Start off with a Vietnamese <a href="https://tasty.co/recipe/banh-mi-meatball-sandwich">Banh Mi Meatball Sandwich</a> and work your way to our hot <a href="https://tasty.co/recipe/fluffy-perfect-pancakes">Canadian pancake breakfast</a>. Enjoy the taste of every culture right from your kitchen!', 'draft_status': 'published', 'facebook_posts': [], 'id': 1893, 'is_shoppable': False, 'keywords': None, 'language': 'eng', 'name': 'Breakfast Around The World', 'promotion': 'full', 'show': [{'id': 17, 'name':

In [3]:
response

<Response [200]>

In [4]:
daikon_recipes = pd.json_normalize(response.json(), "results")
daikon_recipes

approved_at aspect_ratio  ...      brand.name      brand.slug
0    1570199933          1:1  ...             NaN             NaN
1    1582749647          1:1  ...             NaN             NaN
2    1544035794          1:1  ...             NaN             NaN
3    1533335386          1:1  ...             NaN             NaN
4    1528771181          1:1  ...             NaN             NaN
5    1561944458          1:1  ...             NaN             NaN
6    1540495868          1:1  ...             NaN             NaN
7    1553696219          1:1  ...             NaN             NaN
8    1616591279         16:9  ...             NaN             NaN
9    1558026831          1:1  ...             NaN             NaN
10   1568686659          1:1  ...             NaN             NaN
11   1561741987         None  ...             NaN             NaN
12   1599224363          1:1  ...  VisitSingapore  visitsingapore

[13 rows x 76 columns]

In [5]:
import requests
response = requests.get("https://statistics.calpoly.edu/content/directory")

from bs4 import BeautifulSoup
soup = BeautifulSoup(response.content, "html.parser")

results = soup.find(id="block-system-main")
print(results.prettify())

<div class="block block-system" id="block-system-main">
 <div class="field field-name-body field-type-text-with-summary field-label-hidden">
  <div class="field-items">
   <div class="field-item even" property="content:encoded">
    <div>
     Statistics Department
    </div>
    <div>
     California Polytechnic State University
    </div>
    <div>
     Building 25, Room 107D
     <br/>
     San Luis Obispo, CA 93407-0405
     <br/>
     Telephone: (805) 756-2709
    </div>
    <div>
     Email:
     <a href="mailto:statisticsdepartment@calpoly.edu">
      statisticsdepartment@calpoly.edu
     </a>
    </div>
    <h2 id="Summer2024Office_Hours">
     Summer 2024 Office Hours
    </h2>
    <p>
     Office hours for each instructor have been posted below. For additional information, please review your course syllabus and/or visit your course's Canvas page.
    </p>
    <table cellspacing="0" class="directory backwhite" summary="COSAM Dean's Office Staff" width="100%">
     <tbody>
    

In [6]:
tbls = results.find_all("table")
tbls[1]

<table cellspacing="0" class="directory backwhite" height="2667" summary="COSAM Dean's Office Staff" width="596"><tbody><tr><th scope="col" width="18%">
<p>Name</p>
</th>
<th scope="col" width="8%">Office</th>
<th scope="col" width="10%">Phone (805)</th>
<th scope="col" width="20%">Email</th>
<th scope="col" width="20%">
<p>Office HoURS </p>
</th>
</tr><tr><td><a href="/Kelly-Bodwin"><strong>Kelly Bodwin</strong></a></td>
<td><a href="https://maps.calpoly.edu/">25-106</a></td>
<td>756-2450</td>
<td><a href="mailto:kbodwin@calpoly.edu">kbodwin@calpoly.edu</a></td>
<td>
<p>T 12-1:30pm</p>
<p>W 11am-12pm</p>
<p>R 9:30-11am</p>
</td>
</tr><tr><td><a href="/Matt-Carlton"><strong>Matt Carlton</strong></a></td>
<td><a href="https://maps.calpoly.edu/">25-101</a></td>
<td>756-7076</td>
<td><a href="mailto:mcarlton@calpoly.edu">mcarlton@calpoly.edu</a></td>
<td>
<p>M 5:10-6:30pm</p>
<p>W 2:10-3pm</p>
<p>F 11:30am-1pm; Zoom-Course Canvas Page</p>
</td>
</tr><tr><td>
<p><a href="/Beth-Chance"><str

## Scraping an HTML table with Beautiful Soup

Open the URL https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population and scroll down until you see a table of the cities in the U.S. with population over 100,000 (as of Jul 1, 2022). We'll use Beautiful Soup to scrape information from this table.

Read in the HTML from the ULR using the `requests` library.

In [7]:
# YOUR CODE HERE
response = requests.get("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

Use Beautiful Soup to parse this string into a tree called `soup`

In [8]:
# YOUR CODE HERE
from bs4 import BeautifulSoup
soup = BeautifulSoup(response.content, "html.parser")

To find an HTML tag corresponding to a specific element on a webpage, right-click on it and choose "Inspect element". Go to the cities table Wikipedia page and do this now.

You should find that the cities table on the Wikipedia page corresponds to the element

```
<table class="wikitable sortable jquery-tablesorter" style="text-align:center">
```

There are many `<table>` tags on the page.

In [9]:
len(soup.find_all("table"))

10

We can use attributes like `class=` and `style=` to narrow down the list.

In [10]:
len(soup.find_all("table",
                  attrs={
                      "class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                      'style': "text-align:right"}
                  ))

1

At this point, you can manually inspect the tables on the webpage to find that the one we want is the first one (see `[0]` below). We'll store this as `table`.

In [11]:
table = soup.find_all("table",
                  attrs={
                      "class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                      'style': "text-align:right"}
                  )[0]

table

<table class="sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center" style="text-align:right">
<tbody><tr>
<th rowspan="2">City
</th>
<th rowspan="2"><abbr title="State or district">ST</abbr>
</th>
<th rowspan="2">2023<br/>estimate
</th>
<th rowspan="2">2020<br/>census
</th>
<th rowspan="2">Change
</th>
<th colspan="2">2020 land area
</th>
<th colspan="2">2020 density
</th>
<th rowspan="2">Location
</th></tr>
<tr class="static-row-header">
<th>mi<sup>2</sup>
</th>
<th>km<sup>2</sup>
</th>
<th>/ mi<sup>2</sup>
</th>
<th>/ km<sup>2</sup>
</th></tr>
<tr>
<td style="background-color:#cfecec"><i><a href="/wiki/New_York_City" title="New York City">New York</a></i><sup class="reference" id="cite_ref-5"><a href="#cite_note-5"><span class="cite-bracket">[</span>c<span class="cite-bracket">]</span></a></sup>
</td>
<td><a href="/wiki/New_York_(state)" title="New York (state)">NY</a>
</td>
<td style="text-align:right;">8,258,035
</td>
<td style="text-align:right;

**Now you will write code to scrape the information in `table` to create a Pandas data frame with one row for each city and columns for: city, state, population (2022 estimate), and 2020 land area (sq mi).** Refer to the Notes/suggestions below as you write your code. A few Hints are provided further down, but try coding first before looking at the hints.

Notes/suggestions:

- Use as a guide the code from the reading that produced the data frame of Statistics faculty
- Inspect the page source as you write your code
- You will need to write a loop to get the information for all cities, but you might want to try just scraping the info for New York first
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `.get_text(strip = True)` instead of `.text`
- Don't forget to convert to a Pandas Data Frame; it should have 333 rows and 4 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it --- e.g., what is the population density for all cities in CA? --- then you would need to clean the data first (to clean strings and convert to quantitative). (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)

In [12]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [13]:
# initialize an empty list
rows = []

# iterate over all rows in the pop table which is the table that we specified above
for pop in table.find_all("tr")[1:]:

    # Getting all the cells (<td>) in the row.
    cells = pop.find_all("td")

    # Check if cells are found before proceeding
    if cells: # This condition ensures that cells is not empty before accessing the elements

        # The information we need is the text between tags. So we need to look through them

        # Find the the city of the pop table in cell[0]
        # which for most pop is contained in the <strong> tag
        city_tag = cells[0]
        city = city_tag.get_text(strip = True)

        # Find the state of the pop table in cell[1]
        # which for most pop is contained in the <a> tag
        state_tag = cells[1].find("a") or cells[1]
        state = state_tag.text

        # Find the populaiton of the pop table in cell[3]
        # which for most pop is contained in the <a> tag
        population_tag = cells[2].find("a") or cells[2]
        population = population_tag.get_text(strip = True)

        # Find the land area of the pop in cell[2]
        # which for most pop is contained in the <a> tag
        land_area_tag = cells[5].find("a") or cells[5]
        land_area = land_area_tag.get_text(strip = True)

        # Append this data.
        rows.append({
            "city": city,
            "state": state,
            "population": population,
            "land_area": land_area
        })

# visualizing our list that we created
rows

[{'city': 'New York[c]',
  'state': 'NY',
  'population': '8,258,035',
  'land_area': '300.5'},
 {'city': 'Los Angeles',
  'state': 'CA',
  'population': '3,820,914',
  'land_area': '469.5'},
 {'city': 'Chicago',
  'state': 'IL',
  'population': '2,664,452',
  'land_area': '227.7'},
 {'city': 'Houston',
  'state': 'TX',
  'population': '2,314,157',
  'land_area': '640.4'},
 {'city': 'Phoenix',
  'state': 'AZ',
  'population': '1,650,070',
  'land_area': '518.0'},
 {'city': 'Philadelphia[d]',
  'state': 'PA',
  'population': '1,550,542',
  'land_area': '134.4'},
 {'city': 'San Antonio',
  'state': 'TX',
  'population': '1,495,295',
  'land_area': '498.8'},
 {'city': 'San Diego',
  'state': 'CA',
  'population': '1,388,320',
  'land_area': '325.9'},
 {'city': 'Dallas',
  'state': 'TX',
  'population': '1,302,868',
  'land_area': '339.6'},
 {'city': 'Jacksonville[e]',
  'state': 'FL',
  'population': '985,843',
  'land_area': '747.3'},
 {'city': 'Austin',
  'state': 'TX',
  'population': 

Hints:

- Each city is a row in the table; find all the `<tr>` tags to find all the cities
- Look for the `<td>` tag to see table entries within a row
- The rank column is represented by `<th>` tags, rather than `<td>` tags. So within a row, the first (that is, `[0]`) `<td>` tag corresponds to the city name.

## Aside: Scraping an HTML table with Pandas



The Pandas command `read_html` can be used to scrape information from an HTML table on a webpage.

We can call `read_html` on the URL.

In [14]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

ImportError: Missing optional dependency 'lxml'.  Use pip or conda to install lxml.

However, this scrapes all the tables on the webpage, not just the one we want. As with Beautiful Soup, we can narrow the search by specifying the table attributes.

In [ ]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {'class': "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center", 'style': "text-align:right"})

This still returns 3 tables. As we remarked above, the table that we want is the first one (see `[0]` below).

In [15]:
df_cities2 = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {'class': "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center", 'style': "text-align:right"})[0]
df_cities2

ImportError: Missing optional dependency 'lxml'.  Use pip or conda to install lxml.

Wait, that seemed much easier than using Beautiful Soup, and it returned a data frame, and we even got for free some formatting like removing the commas from the population! Why didn't we just use `read_html` in the first place? It's true the `read_html` works well when scraping information from an HTML *table*. Unfortunately, you often want to scrape information from a webpage that isn't conveniently stored in an HTML table, in which case `read_html` won't work. (It only searches for `<table>`, `<th>`, `<tr>`, and `<td>` tags, but there are many other HTML tags.) Though Beautiful Soup is not as simple as `read_html`, it is more flexible and thus more widely applicable.

## Scraping information that is NOT in a `<table>` with Beautiful Soup

The Cal Poly course catalog http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory contains a list of courses offered by the Statistics department. **You will scrape this website to obtain a Pandas data frame with one row for each DATA or STAT course and two columns: course name and number (e.g, DATA 301. Introduction to Data Science) and term typically offered (e.g., Term Typically Offered: F, W, SP).**

Note: Pandas `read_html` is not help here since the courses are not stored in a `<table>.`

In [ ]:
pd.read_html("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")


Notes/suggestions:


- Inspect the page source as you write your code
- The courses are not stored in a `<table>`. How are they stored?
- You will need to write a loop to get the information for all courses, but you might want to try just scraping the info for DATA 100 first
- What kind of tag is the course name stored in? What is the `class` of the tag?
- What kind of tag is the quarter(s) the course is offered stored in? What is the `class` of the tag? Is this the only tag of this type with the class? How will you get the one you want?
- You don't have to remove the number of units (e.g., 4 units) from the course name and number, but you can try it if you want
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `get_text(strip = True)` instead of `text`
- Don't forget to convert to a Pandas Data Frame; it should have 74 rows and 2 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it then you might need to clean the data first. (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)



In [16]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [17]:
#making the request for the page
res = requests.get("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")

#making a soup with that request from the page with parsing
soup2 = BeautifulSoup(res.content, 'html.parser')

In [18]:
#visualizing one of the divs so we can see what attributes to look for
div = soup2.find_all("div", attrs={"class": "courseblock"})[0]
div

<div class="courseblock">
<p class="courseblocktitle"><strong>DATA 100. Data Science for All I.
<span class="courseblockhours">4 units
</span></strong></p><div class="noindent courseextendedwrap">
<p class="noindent">Term Typically Offered: F, W, SP</p><p class="noindent">2020-21 or later catalog: GE Area B4</p><p class="noindent">2019-20 or earlier catalog: GE Area B4</p><p>Prerequisite: <a class="bubblelink code" href="/search/?P=MATH%20115" onclick="return showCourse(this, 'MATH 115');" title="MATH 115">MATH 115</a>, <a class="bubblelink code" href="/search/?P=MATH%20116" onclick="return showCourse(this, 'MATH 116');" title="MATH 116">MATH 116</a>, <a class="bubblelink code" href="/search/?P=MATH%20118" onclick="return showCourse(this, 'MATH 118');" title="MATH 118">MATH 118</a>, or Appropriate Math Placement Level.</p></div>
<div class="courseblockdesc">
<p>Basic approaches for answering questions using data.  Emphasis on working with tabular data in spreadsheet software to provide

In [19]:
#creating an empty list to put our scraped items in
rows1 = []

#iterating through the div to get the information that we need
for i in soup2.find_all("div", attrs={"class": "courseblock"})[0:]:

  #looking through the div finding p with specified attributes to find classes
  class_chunk = i.find("p", attrs={"class": "courseblocktitle"})
  dummy1 = class_chunk.get_text(strip = True)

  #looking through the div finding p with specified attributes to find terms
  cells_term = i.find("p", attrs={"class": "noindent"})
  dummy2 = cells_term.get_text(strip = True)

  #filling up the list with the web scraped data
  rows1.append({"class": dummy1, "term": dummy2})

In [20]:
#visualizing the list
rows1

[{'class': 'DATA\xa0100. Data Science for All I.4 units',
  'term': 'Term Typically Offered: F, W, SP'},
 {'class': 'DATA\xa0301. Introduction to Data Science.4 units',
  'term': 'Term Typically Offered: F, W, SP'},
 {'class': 'DATA\xa0401. Data Science Process and Ethics.3 units',
  'term': 'Term Typically Offered: F'},
 {'class': 'DATA\xa0402. Mathematical Foundations of Data Science.3 units',
  'term': 'Term Typically Offered: F'},
 {'class': 'DATA\xa0403. Data Science Projects Laboratory.1 unit',
  'term': 'Term Typically Offered: F'},
 {'class': 'DATA\xa0441. Bioinformatics Capstone I.2 units',
  'term': 'Term Typically Offered: W'},
 {'class': 'DATA\xa0442. Bioinformatics Capstone II.2 units',
  'term': 'Term Typically Offered: SP'},
 {'class': 'DATA\xa0451. Data Science Capstone I.2 units',
  'term': 'Term Typically Offered: W'},
 {'class': 'DATA\xa0452. Data Science Capstone II.2 units',
  'term': 'Term Typically Offered: SP'},
 {'class': 'DATA\xa0472. Data Science Seminar.1 un

In [21]:
#creating dataframe from list
df_class = pd.DataFrame(rows1)
df_class

class                              term
0            DATA 100. Data Science for All I.4 units  Term Typically Offered: F, W, SP
1      DATA 301. Introduction to Data Science.4 units  Term Typically Offered: F, W, SP
2   DATA 401. Data Science Process and Ethics.3 units         Term Typically Offered: F
3   DATA 402. Mathematical Foundations of Data Sci...         Term Typically Offered: F
4   DATA 403. Data Science Projects Laboratory.1 unit         Term Typically Offered: F
..                                                ...                               ...
69      STAT 551. Statistical Learning with R.4 units         Term Typically Offered: F
70    STAT 566. Graduate Consulting Practicum.2 units        Term Typically Offered: SP
71       STAT 570. Selected Advanced Topics.1-4 units       Term Typically Offered: TBD
72    STAT 590. Graduate Seminar in Statistics.1 unit  Term Typically Offered: F, W, SP
73                         STAT 599. Thesis.1-4 units       Term Typically Offered: TBD

[74 rows x 2 columns]

Hints:

- Each course is represented by a `<div>` with `class=courseblock`, so you can find all the courses with `soup.find_all("div", {"class": "courseblock"})`
- The course name is in a `<p>` tag with `class=courseblocktitle`, inside a `<strong>` tag. (Though I don't think we need to find the strong tag here.)
- The term typically offered is in `<p>` tag with `class=noindent`. However, there are several tags with this class; term typically offered is the first one.
- If you want to use Beautiful Soup to remove the course units (e.g., 4 units), find the `<span>` tag within the course name tag and `.extract()` this span tag